proxy/src/transparency/server.rs

```diff
@@ -0,0 +1,158 @@
+use std::fmt;
+use std::net::SocketAddr;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use futures::Future;
+use http;
+use hyper;
+use tokio_core::reactor::Handle;
+use tower::NewService;
+use tower_h2;
+
+use control;
+use connection::Connection;
+use ctx::Proxy as ProxyCtx;
+use ctx::transport::{Server as ServerCtx};
+use telemetry::Sensors;
+use transport::GetOriginalDst;
+use super::glue::{HttpBody, HttpBodyNewSvc, HyperServerSvc};
+use super::protocol::Protocol;
+use super::tcp;
+
+pub struct Server<S: NewService, B: tower_h2::Body, G>
+where
+    S: NewService<Request=http::Request<HttpBody>>,
+    S::Future: 'static,
+{
+    executor: Handle,
+    get_orig_dst: G,
+    h1: hyper::server::Http,
+    h2: tower_h2::Server<HttpBodyNewSvc<S>, Handle, B>,
+    listen_addr: SocketAddr,
+    new_service: S,
+    proxy_ctx: Arc<ProxyCtx>,
+    sensors: Sensors,
+    tcp: tcp::Proxy,
+}
+
+impl<S, B, G> Server<S, B, G>
+where
+    S: NewService<
+        Request = http::Request<HttpBody>,
+        Response = http::Response<B>
+    > + Clone + 'static,
+    S::Future: 'static,
+    S::Error: fmt::Debug,
+    B: tower_h2::Body + 'static,
+    G: GetOriginalDst,
+{
+    pub fn new(
+        listen_addr: SocketAddr,
+        proxy_ctx: Arc<ProxyCtx>,
+        sensors: Sensors,
+        get_orig_dst: G,
+        stack: S,
+        tcp_connect_timeout: Duration,
+        executor: Handle,
+    ) -> Self {
+        let recv_body_svc = HttpBodyNewSvc::new(stack.clone());
+        let tcp = tcp::Proxy::new(tcp_connect_timeout, sensors.clone(), &executor);
+        Server {
+            executor: executor.clone(),
+            get_orig_dst,
+            h1: hyper::server::Http::new(),
+            h2: tower_h2::Server::new(recv_body_svc, Default::default(), executor),
+            listen_addr,
+            new_service: stack,
+            proxy_ctx,
+            sensors,
+            tcp,
+        }
+    }
+
+    pub fn serve(&self, connection: Connection, remote_addr: SocketAddr) {
```

Comment:
This would benefit from a comment outlining the general story of what the server does.